### PR TITLE
db fetch with guti and amfUeNgapId

### PR DESCRIPTION
**Pull request by badhrinathpa**

### CODE REVIEW

File: util/path.go [hunk 1]
```diff
@@ -3,7 +3,8 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 
-//+build !debug
+//go:build !debug
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply util/path.go:6]: go fmt is adding this change.

File: context/amf_ue.go [hunk 2]
```diff
@@ -60,144 +63,265 @@ const (
...
+		ULCount     uint32                       `json:"ulCount"`
+		DLCount     uint32                       `json:"dlCount"`
+		RanUeNgapId int64                        `json:"ranUeNgapId"`
+		AmfUeNgapId int64                        `json:"amfUeNgapId"`
+		*Alias
+	}{
+		Alias: (*Alias)(ue),
+	}
+	if err := json.Unmarshal(data, &aux); err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply context/amf_ue.go:301]: done

File: context/db.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+		logger.ContextLog.Errorf("Create index failed on Tmsi field.")
+	}
+
+	_, err = MongoDBLibrary.CreateIndex(AmfUeDataColl, "amfUeNgapId")
+	if err != nil {
+		logger.ContextLog.Errorf("Create index failed on AmfUeNgapID field.")
+	}
+
+	_, err = MongoDBLibrary.CreateIndex(AmfUeDataColl, "ranUeNgapId")
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply context/db.go:103]: done

File: context/db.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+		logger.ContextLog.Errorf("Create index failed on RanUeNgapID field.")
+	}
+}
+
+func ToBsonM(data *AmfUe) (ret bson.M) {
+	tmp, err := json.Marshal(data)
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply context/db.go:109]: TBD

File: context/db.go [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+	ue.RanUe[models.AccessType__3_GPP_ACCESS].AmfUe = ue
+	ue.EventChannel = nil
+	ue.NASLog = logger.NasLog.WithField(logger.FieldAmfUeNgapID, fmt.Sprintf("AMF_UE_NGAP_ID:%d", ue.RanUe[models.AccessType__3_GPP_ACCESS].AmfUeNgapId))
+	ue.GmmLog = logger.GmmLog.WithField(logger.FieldAmfUeNgapID, fmt.Sprintf("AMF_UE_NGAP_ID:%d", ue.RanUe[models.AccessType__3_GPP_ACCESS].AmfUeNgapId))
+	ue.TxLog = logger.GmmLog.WithField(logger.FieldAmfUeNgapID, fmt.Sprintf("AMF_UE_NGAP_ID:%d", ue.RanUe[models.AccessType__3_GPP_ACCESS].AmfUeNgapId))
+	return ue.RanUe[models.AccessType__3_GPP_ACCESS]
+}
+
+func DbFetchUeByGuti(guti string) (ue *AmfUe, ok bool) {
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply context/db.go:213]: done

File: context/amf_ue.go [hunk 6]
```diff
@@ -60,144 +63,265 @@ const (
...
+	})
+
+	return json.Marshal(&struct {
+		State       map[models.AccessType]string `json:"state"`
+		SmCtxList   map[string]SmContext         `json:"smCtxList"`
+		N1N2Message N1N2Message                  `json:"n1n2Msg"`
+		ULCount     uint32                       `json:"ulCount"`
+		DLCount     uint32                       `json:"dlCount"`
+		RanUeNgapId int64                        `json:"ranUeNgapId"`
+		AmfUeNgapId int64                        `json:"amfUeNgapId"`
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply context/amf_ue.go]: done

File: context/amf_ran.go [hunk 7]
```diff
@@ -101,7 +102,7 @@ func (ran *AmfRan) RanUeFindByRanUeNgapID(ranUeNgapID int64) *RanUe {
 			return ranUe
 		}
 	}
-	return nil
+	return DbFetchRanUeByRanUeNgapID(ranUeNgapID)
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply context/amf_ran.go]: done changes to use db filter as combination of ranuengapid and ranremoteIP

File: context/context.go [hunk 8]
```diff
@@ -388,15 +391,26 @@ func (context *AMFContext) InPlmnSupportList(snssai models.Snssai) bool {
...
 func (context *AMFContext) AmfUeFindByGuti(guti string) (ue *AmfUe, ok bool) {
-	context.UePool.Range(func(key, value interface{}) bool {
-		candidate := value.(*AmfUe)
-		if ok = (candidate.Guti == guti); ok {
-			ue = candidate
-			return false
-		}
-		return true
-	})
+	if context.EnableDbStore {
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply context/context.go]: done

File: context/transaction.go [hunk 9]
```diff
@@ -41,6 +41,10 @@ func (tx *EventChannel) Start() {
 		case msg := <-tx.Message:
 			switch msg.(type) {
 			case NasMsg:
+				if tx.NasHandler == nil {
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply context/transaction.go]: removed check. not needed

File: context/amf_ue.go [hunk 10]
```diff
@@ -60,144 +63,265 @@ const (
 )
 
 type AmfUe struct {
-	Mutex sync.Mutex
+	//Mutex sync.Mutex `json:"mutex,omitempty" yaml:"mutex" bson:"mutex,omitempty"`
+	Mutex sync.Mutex `json:"-" yaml:"mutex" bson:"-"`
 	/* the AMF which serving this AmfUe now */
-	servingAMF *AMFContext // never nil
+	servingAMF *AMFContext `json:"servingAMF,omitempty" yaml:"servingAMF" bson:"servingAMF,omitempty"` // never nil
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply context/amf_ue.go]: done

File: context/context.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply context/context.go]: done

File: context/context.go [hunk 11]
```diff
@@ -413,10 +427,14 @@ func (context *AMFContext) AmfUeFindByPolicyAssociationID(polAssoId string) (ue
 }
 
 func (context *AMFContext) RanUeFindByAmfUeNgapID(amfUeNgapID int64) *RanUe {
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply context/context.go:459]: TBD

File: context/context.go [hunk 12]
```diff
@@ -413,10 +427,14 @@ func (context *AMFContext) AmfUeFindByPolicyAssociationID(polAssoId string) (ue
...
-	if value, ok := context.RanUePool.Load(amfUeNgapID); ok {
-		return value.(*RanUe)
+	if context.EnableDbStore {
+		return DbFetchRanUeByAmfUeNgapID(amfUeNgapID)
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply context/context.go]: done

File: context/db.go [hunk 13]
```diff
@@ -0,0 +1,176 @@
+// SPDX-FileCopyrightText: 2021 Open Networking Foundation <info@opennetworking.org>
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply context/db.go]: done

File: ngap/handler.go [hunk 14]
```diff
@@ -101,7 +105,11 @@ func FetchRanUeContext(ran *context.AmfRan, message *ngapType.NGAPPDU) *context.
 						}
 						ranUe.Log.Warnf("Known UE [GUTI: %s]", guti)
 						amfUe.AttachRanUe(ranUe)
+					} else {
+						ran.Log.Errorf("find by guti failed")
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply ngap/handler.go]: done

File: ngap/handler.go [hunk 15]
```diff
@@ -1343,21 +1357,27 @@ func HandleInitialUEMessage(ran *context.AmfRan, message *ngapType.NGAPPDU) {
...
-		triggeringMessage := ngapType.TriggeringMessagePresentInitiatingMessage
-		procedureCriticality := ngapType.CriticalityPresentIgnore
-		criticalityDiagnostics := buildCriticalityDiagnostics(&procedureCode, &triggeringMessage, &procedureCriticality,
-			nil)
-		cause := ngapType.Cause{
-			Present: ngapType.CausePresentProtocol,
-			Protocol: &ngapType.CauseProtocol{
-				Value: ngapType.CauseProtocolPresentMessageNotCompatibleWithReceiverState,
-			},
+	if !(amfSelf.EnableSctpLb && amfSelf.EnableDbStore) {
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply ngap/handler.go]: done

File: producer/n1n2message.go [hunk 16]
```diff
@@ -266,6 +267,7 @@ func N1N2MessageTransferProcedure(ueContextID string, reqUri string,
 				}
 				n1n2MessageTransferRspData = new(models.N1N2MessageTransferRspData)
 				n1n2MessageTransferRspData.Cause = models.N1N2MessageTransferCause_N1_N2_TRANSFER_INITIATED
+				context.StoreContextInDB(ue)
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply producer/n1n2message.go]: not needed. can be added later.

File: producer/n1n2message.go [hunk 17]
```diff
@@ -274,6 +276,7 @@ func N1N2MessageTransferProcedure(ueContextID string, reqUri string,
 				ngap_message.SendPDUSessionResourceModifyRequest(ue.RanUe[anType], list)
 				n1n2MessageTransferRspData = new(models.N1N2MessageTransferRspData)
 				n1n2MessageTransferRspData.Cause = models.N1N2MessageTransferCause_N1_N2_TRANSFER_INITIATED
+				context.StoreContextInDB(ue)
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply producer/n1n2message.go]: commented

File: ngap/dispatcher.go [hunk 18]
```diff
@@ -97,14 +107,19 @@ func Dispatch(conn net.Conn, msg []byte) {
 	if ranUe != nil && ranUe.AmfUe != nil {
 		if ranUe.AmfUe.EventChannel == nil {
 			ran.Log.Errorf("AmfUe EventChannel is not exist")
-			return
+			ran.Log.Errorf("Creating new AmfUe EventChannel")
+			ranUe.AmfUe.EventChannel = ranUe.AmfUe.NewEventChannel()
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply ngap/dispatcher.go]: done

File: factory/config.go [hunk 19]
```diff
@@ -40,6 +40,9 @@ const (
 
 type Configuration struct {
 	AmfName                         string                    `yaml:"amfName,omitempty"`
+	AmfDBName                       string                    `yaml:"amfDBName,omitempty"`
+	MongoDBName                     string                    `yaml:"MongoDBName"`
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply factory/config.go]: done

File: context/db.go [hunk 20]
```diff
@@ -0,0 +1,227 @@
...
+	defer dbMutex.Unlock()
+	if *generator == nil {
+		logger.ContextLog.Infof("generator null. fetch offset from db")
+		val := MongoDBLibrary.GetUniqueIdentity(idName)
+		// Mongodb returns value starting from 1.
+		// Limiting users to 5000 per instance.
+		// TODO : Make this value configurable.
+		//        Later this value can be used to trigger
+		//        creation of new instance
+		minVal := int64((val-1)*5000 + 1)
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply context/db.go]: done

File: context/db.go [hunk 21]
```diff
@@ -0,0 +1,227 @@
...
+		maxVal := int64(minVal + 5000)
+		*generator = idgenerator.NewGenerator(minVal, maxVal)
+	}
+
+	val, err := (*generator).Allocate()
+	if err != nil {
+		logger.ContextLog.Infof("Max Users reached for Instance")
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply context/db.go]: done

File: context/db.go [hunk 22]
```diff
@@ -0,0 +1,227 @@
...
+	if err != nil {
+		logger.ContextLog.Errorf("Create index failed on Tmsi field.")
+	}
+
+	_, err = MongoDBLibrary.CreateIndex(AmfUeDataColl, "customFieldsAmfUe.amfUeNgapId")
+	if err != nil {
+		logger.ContextLog.Errorf("Create index failed on AmfUeNgapID field.")
+	}
+
+	_, err = MongoDBLibrary.CreateIndex(AmfUeDataColl, "customFieldsAmfUe.ranUeNgapId")
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply context/db.go]: TBD. commented index creation for now

File: ngap/dispatcher.go [hunk 23]
```diff
@@ -53,18 +54,20 @@ func DispatchLb(remoteAddr string, msg []byte, Amf2RanMsgChan chan *sdcoreAmfSer
...
 		ranUe.AmfUe.TxLog.Infof("Uecontext found. queuing ngap message to uechannel")
 		ranUe.AmfUe.EventChannel.UpdateNgapHandler(NgapMsgHandler)
 		ngapMsg := context.NgapMsg{
 			Ran:     ran,
 			NgapMsg: pdu,
 		}
+
+		globalRANNodeID := ranUe.Ran.RanId
+		ranUe.Ran = ran
+		ranUe.Ran.RanId = globalRANNodeID
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply ngap/dispatcher.go]: not needed. removed

File: ngap/dispatcher.go [hunk 24]
```diff
@@ -53,18 +54,20 @@ func DispatchLb(remoteAddr string, msg []byte, Amf2RanMsgChan chan *sdcoreAmfSer
...
 		ranUe.AmfUe.EventChannel.SubmitMessage(ngapMsg)
 	} else {
+		fmt.Println("ranUe nil :  ", ran.RanID())
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply ngap/dispatcher.go:67]: removed

File: context/db.go [hunk 25]
```diff
@@ -0,0 +1,176 @@
...
+	"github.com/omec-project/amf/logger"
+	"github.com/omec-project/openapi/models"
+)
+
+const (
+	AmfUeDataColl = "amf.data.amfState"
+)
+
+func SetupAmfCollection() {
+	MongoDBLibrary.SetMongoDB("sdcore", "mongodb://mongodb")
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply context/db.go]: done

File: context/db.go [hunk 26]
```diff
@@ -0,0 +1,176 @@
...
+	"fmt"
+	"github.com/omec-project/MongoDBLibrary"
+	"go.mongodb.org/mongo-driver/bson"
+
+	"github.com/omec-project/amf/logger"
+	"github.com/omec-project/openapi/models"
+)
+
+const (
+	AmfUeDataColl = "amf.data.amfState"
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply context/db.go]: done

File: context/db.go [hunk 27]
```diff
@@ -0,0 +1,176 @@
...
+	if err != nil {
+		logger.ContextLog.Errorf("Create index failed on Guti field.")
+	}
+
+	_, err = MongoDBLibrary.CreateIndex(AmfUeDataColl, "tmsi")
+	if err != nil {
+		logger.ContextLog.Errorf("Create index failed on Tmsi field.")
+	}
+
+	_, err = MongoDBLibrary.CreateIndex(AmfUeDataColl, "amfUeNgapId")
```
[the remark this answers is not in the document]
Comment (badhrinathpa) [reply context/db.go]: done

**Review by thakurajayL — APPROVED** [review]

**Pull request merged**